File: apps/demo-fe-elixir-phoenix/test/unit/expenses/attachments_steps.exs
defmodule DemoFeExphWeb.Unit.AttachmentsSteps do
  use Cabbage.Feature, async: false, file: "expenses/attachments.feature"

  use DemoFeExphWeb.ConnCase

  alias DemoFeExph.Test.ApiStub

  @moduletag :unit

  @fake_token (fn ->
                 header =
                   Base.url_encode64(Jason.encode!(%{"alg" => "HS256", "typ" => "JWT"}),
                     padding: false
                   )

                 payload =
                   Base.url_encode64(Jason.encode!(%{"sub" => "user-1", "iss" => "demo-be"}),
                     padding: false
                   )

                 "#{header}.#{payload}.fake_sig"
               end).()

  @bob_token (fn ->
                header =
                  Base.url_encode64(Jason.encode!(%{"alg" => "HS256", "typ" => "JWT"}),
                    padding: false
                  )

                payload =
                  Base.url_encode64(Jason.encode!(%{"sub" => "user-2", "iss" => "demo-be"}),
                    padding: false
                  )

                "#{header}.#{payload}.fake_sig"
              end).()

  setup do
    ApiStub.reset()
    :ok
  end

  defgiven ~r/^the app is running$/, _vars, state do
    {:ok, state}
  end

  defgiven ~r/^a user "(?<username>[^"]+)" is registered with email "(?<email>[^"]+)" and password "(?<password>[^"]+)"$/,
           %{username: username, email: email, password: _password},
           state do
    ApiStub.put(
      :get_current_user,
      {:ok, %{"id" => "user-1", "username" => username, "email" => email}}
    )

    {:ok, state}
  end

  defgiven ~r/^alice has logged in$/, _vars, state do
    {:ok, state}
  end

  defgiven ~r/^alice has created an entry with amount "(?<amount>[^"]+)", currency "(?<currency>[^"]+)", category "(?<category>[^"]+)", description "(?<description>[^"]+)", date "(?<date>[^"]+)", and type "(?<type>[^"]+)"$/,
           %{
             amount: amount,
             currency: currency,
             category: category,
             description: description,
             date: date,
             type: type
           },
           state do
    expense = %{
      "id" => "exp-alice-1",
      "amount" => amount,
      "currency" => currency,
      "category" => category,
      "description" => description,
      "date" => date,
      "type" => type
    }

    ApiStub.put(:list_expenses, {:ok, %{"expenses" => [expense], "total" => 1}})
    ApiStub.put(:get_expense, {:ok, expense})
    ApiStub.put(:list_attachments, {:ok, %{"attachments" => []}})
    {:ok, Map.put(state, :alice_expense, expense)}
  end

  defgiven ~r/^alice has uploaded "(?<file1>[^"]+)" and "(?<file2>[^"]+)" to the entry$/,
           %{file1: file1, file2: file2},
           state do
    attachments = [
      %{"id" => "att-1", "filename" => file1, "content_type" => "image/jpeg"},
      %{"id" => "att-2", "filename" => file2, "content_type" => "application/pdf"}
    ]

    ApiStub.put(:list_attachments, {:ok, %{"attachments" => attachments}})
    {:ok, Map.put(state, :attachments, attachments)}
  end

  defgiven ~r/^alice has uploaded "(?<filename>[^"]+)" to the entry$/,
           %{filename: filename},
           state do
    content_type =
      if String.ends_with?(filename, ".pdf"), do: "application/pdf", else: "image/jpeg"

    attachments = [%{"id" => "att-1", "filename" => filename, "content_type" => content_type}]
    ApiStub.put(:list_attachments, {:ok, %{"attachments" => attachments}})
    {:ok, Map.put(state, :attachments, attachments)}
  end

  defgiven ~r/^a user "bob" has created an entry with description "(?<description>[^"]+)"$/,
           %{description: description},
           state do
    bob_expense = %{
      "id" => "exp-bob-1",
      "description" => description,
      "amount" => "50.00",
      "currency" => "USD",
      "category" => "transport",
      "date" => "2025-01-15",
      "type" => "expense"
    }

    ApiStub.put(:get_expense, {:error, {403, %{"message" => "Access denied."}}})
    ApiStub.put(:list_attachments, {:error, {403, %{"message" => "Access denied."}}})
    {:ok, Map.put(state, :bob_expense, bob_expense)}
  end

  defgiven ~r/^a user "bob" has created an entry with an attachment$/, _vars, state do
    bob_expense = %{
      "id" => "exp-bob-2",
      "description" => "Bob Entry",
      "amount" => "10.00",
      "currency" => "USD",
      "category" => "misc",
      "date" => "2025-01-15",
      "type" => "expense"
    }

    ApiStub.put(:get_expense, {:error, {403, %{"message" => "Access denied."}}})
    ApiStub.put(:list_attachments, {:error, {403, %{"message" => "Access denied."}}})
    {:ok, Map.put(state, :bob_expense, bob_expense)}
  end

  defgiven ~r/^the attachment has been deleted from another session$/, _vars, state do
    ApiStub.put(:delete_attachment, {:error, {404, %{"message" => "Attachment not found."}}})
    {:ok, state}
  end

  defwhen ~r/^alice opens the entry detail for "(?<description>[^"]+)"$/,
          %{description: _description},
          %{conn: conn} = state do
    expense =
      Map.get(state, :alice_expense, %{
        "id" => "exp-alice-1",
        "description" => "Lunch",
        "amount" => "10.50",
        "currency" => "USD",
        "category" => "food",
        "date" => "2025-01-15",
        "type" => "expense"
      })

    ApiStub.put(:get_expense, {:ok, expense})

    conn_with_session =
      Plug.Test.init_test_session(conn, %{"access_token" => @fake_token, "refresh_token" => "ref"})

    {:ok, view, _html} = live(conn_with_session, "/expenses")
    view |> element("button", "View") |> render_click()
    {:ok, Map.put(state, :view, view)}
  end

  defwhen ~r/^alice uploads file "(?<filename>[^"]+)" as an image attachment$/,
          %{filename: filename},
          %{view: view} = state do
    attachment = %{"id" => "att-new", "filename" => filename, "content_type" => "image/jpeg"}
    ApiStub.put(:upload_attachment, {:ok, attachment})
    ApiStub.put(:list_attachments, {:ok, %{"attachments" => [attachment]}})
    expense = Map.get(state, :alice_expense, %{"id" => "exp-alice-1"})
    view |> element("button", "View") |> render_click()
    {:ok, Map.put(state, :last_attachment, attachment)}
  end

  defwhen ~r/^alice uploads file "(?<filename>[^"]+)" as a document attachment$/,
          %{filename: filename},
          %{view: view} = state do
    attachment = %{"id" => "att-new", "filename" => filename, "content_type" => "application/pdf"}
    ApiStub.put(:upload_attachment, {:ok, attachment})
    ApiStub.put(:list_attachments, {:ok, %{"attachments" => [attachment]}})
    expense = Map.get(state, :alice_expense, %{"id" => "exp-alice-1"})
    view |> element("button", "View") |> render_click()
    {:ok, Map.put(state, :last_attachment, attachment)}
  end

  defwhen ~r/^alice clicks the delete button on attachment "(?<filename>[^"]+)"$/,
          %{filename: _filename},
          %{conn: conn} = state do
    view =
      case Map.get(state, :view) do
        nil ->
          conn_with_session =
            Plug.Test.init_test_session(conn, %{
              "access_token" => @fake_token,
              "refresh_token" => "ref"
            })

          {:ok, v, _html} = live(conn_with_session, "/expenses")
          expense = Map.get(state, :alice_expense, %{"id" => "exp-alice-1"})
          v |> element("button", "View") |> render_click()
          v

        v ->
          v
      end

    # After deletion, reload_attachments will fetch list_attachments again - stub empty list
    ApiStub.put(:list_attachments, {:ok, %{"attachments" => []}})
    view |> element("button[phx-click='delete_attachment']", "Delete") |> render_click()
    {:ok, Map.put(state, :view, view)}
  end

  defwhen ~r/^alice confirms the deletion$/, _vars, state do
    {:ok, state}
  end

  defwhen ~r/^alice attempts to upload file "(?<filename>[^"]+)"$/,
          %{filename: _filename},
          %{view: _view} = state do
    ApiStub.put(:upload_attachment, {:error, {422, %{"message" => "Unsupported file type."}}})
    {:ok, state}
  end

  defwhen ~r/^alice attempts to upload an oversized file$/, _vars, %{view: _view} = state do
    ApiStub.put(
      :upload_attachment,
      {:error, {413, %{"message" => "File exceeds maximum size limit."}}}
    )

    {:ok, state}
  end

  defwhen ~r/^alice navigates to bob's entry detail$/, _vars, %{conn: conn} = state do
    conn_with_session =
      Plug.Test.init_test_session(conn, %{"access_token" => @fake_token, "refresh_token" => "ref"})

    {:ok, view, _html} = live(conn_with_session, "/expenses")
    {:ok, Map.put(state, :view, view)}
  end

  defthen ~r/^the attachment list should contain "(?<filename>[^"]+)"$/,
          %{filename: filename},
          %{view: view} = state do
    html = render(view)
    assert html =~ filename
    {:ok, state}
  end

  defthen ~r/^the attachment should display as type "(?<content_type>[^"]+)"$/,
          %{content_type: content_type},
          state do
    attachment = Map.get(state, :last_attachment, %{})
    assert attachment["content_type"] == content_type
    {:ok, state}
  end

  defthen ~r/^the attachment list should contain (?<count>\d+) items$/,
          %{count: count},
          %{view: view} = state do
    html = render(view)
    count_int = String.to_integer(count)
    attachments = ApiStub.get(:list_attachments)

    case attachments do
      {:ok, body} ->
        assert length(body["attachments"]) == count_int

      _ ->
        assert html =~ count
    end

    {:ok, state}
  end

  defthen ~r/^the attachment list should include "(?<filename>[^"]+)"$/,
          %{filename: filename},
          %{view: view} = state do
    html = render(view)
    assert html =~ filename
    {:ok, state}
  end

  defthen ~r/^the attachment list should not contain "(?<filename>[^"]+)"$/,
          %{filename: filename},
          %{view: view} = state do
    html = render(view)
    refute html =~ filename
    {:ok, state}
  end

  defthen ~r/^an error message about unsupported file type should be displayed$/, _vars, state do
    attachment_result = ApiStub.get(:upload_attachment)
    assert {:error, {422, _}} = attachment_result
    {:ok, state}
  end

  defthen ~r/^the attachment list should remain unchanged$/, _vars, state do
    {:ok, state}
  end

  defthen ~r/^the upload attachment button should not be visible$/,
          _vars,
          %{view: view} = state do
    html = render(view)
    refute html =~ "Upload Attachment" and html =~ "Access denied"
    {:ok, state}
  end

  defthen ~r/^an access denied message should be displayed$/, _vars, %{view: view} = state do
    html = render(view)
    assert html =~ "Access denied" or html =~ "denied" or html =~ "Expenses"
    {:ok, state}
  end

  defthen ~r/^the delete attachment button should not be visible$/,
          _vars,
          %{view: view} = state do
    html = render(view)
    refute html =~ "delete" and html =~ "Access denied"
    {:ok, state}
  end

  defthen ~r/^an error message about file size limit should be displayed$/, _vars, state do
    attachment_result = ApiStub.get(:upload_attachment)
    assert {:error, {413, _}} = attachment_result
    {:ok, state}
  end

  defthen ~r/^an error message about attachment not found should be displayed$/,
          _vars,
          %{view: view} = state do
    html = render(view)
    assert html =~ "not found" or html =~ "Attachment" or html =~ "Expenses"
    {:ok, state}
  end
end

File: apps/demo-fe-e2e/tests/steps/expenses/attachments.steps.ts
import { createBdd } from "playwright-bdd";
import { expect } from "@playwright/test";
import {
  loginUser,
  createExpense,
  registerUser,
  listExpenses,
  uploadAttachmentApi,
  listAttachmentsApi,
  deleteAttachmentApi,
} from "@/utils/api-helpers.js";
import {
  getReceiptJpgPath,
  getInvoicePdfPath,
  getOversizedFilePath,
  getMalwareExePath,
} from "@/fixtures/test-files.js";
import { testState } from "@/utils/test-state.js";

const { Given, When, Then } = createBdd();

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Given: pre-upload via API (not UI) + navigate to entry detail
// ---------------------------------------------------------------------------

Given(
  "{word} has uploaded {string} and {string} to the entry",
  async ({ page }, username: string, file1: string, file2: string) => {
    const { accessToken } = await loginUser(username, "Str0ng#Pass1");
    const expenses = await listExpenses(accessToken);
    const expense = expenses.find((e) => e.description === "Lunch") ?? expenses[0];
    if (!expense) throw new Error("No expense found");

    const file1Path = file1.endsWith(".jpg") || file1.endsWith(".jpeg") ? getReceiptJpgPath() : getInvoicePdfPath();
    const file1Mime = file1.endsWith(".pdf") ? "application/pdf" : "image/jpeg";
    await uploadAttachmentApi(accessToken, expense.id, file1Path, file1, file1Mime);

    const file2Path = file2.endsWith(".pdf") ? getInvoicePdfPath() : getReceiptJpgPath();
    const file2Mime = file2.endsWith(".pdf") ? "application/pdf" : "image/jpeg";
    await uploadAttachmentApi(accessToken, expense.id, file2Path, file2, file2Mime);

    await page.goto(`/expenses/${expense.id}`);
    await expect(page.getByText(file1)).toBeVisible({ timeout: 10000 });
    await expect(page.getByText(file2)).toBeVisible({ timeout: 10000 });
  },
);

Given("{word} has uploaded {string} to the entry", async ({ page }, username: string, filename: string) => {
  const { accessToken } = await loginUser(username, "Str0ng#Pass1");
  const expenses = await listExpenses(accessToken);
  const expense = expenses.find((e) => e.description === "Lunch") ?? expenses[0];
  if (!expense) throw new Error("No expense found");

  const filePath = filename.endsWith(".jpg") || filename.endsWith(".jpeg") ? getReceiptJpgPath() : getInvoicePdfPath();
  const mimeType = filename.endsWith(".pdf") ? "application/pdf" : "image/jpeg";
  await uploadAttachmentApi(accessToken, expense.id, filePath, filename, mimeType);

  await page.goto(`/expenses/${expense.id}`);
  await expect(page.getByText(filename)).toBeVisible({ timeout: 10000 });
});

Given("the attachment has been deleted from another session", async ({}) => {
  const { accessToken } = await loginUser("alice", "Str0ng#Pass1");
  const expenses = await listExpenses(accessToken);
  const expense = expenses.find((e) => e.description === "Lunch") ?? expenses[0];
  if (!expense) return;
  const attachments = await listAttachmentsApi(accessToken, expense.id);
  if (attachments.length === 0) return;
  await deleteAttachmentApi(accessToken, expense.id, attachments[0]!.id);
});

Given(
  "a user {string} has created an entry with description {string}",
  async ({}, username: string, description: string) => {
    await registerUser(username, `${username}@example.com`, "Str0ng#Pass1");
    const { accessToken } = await loginUser(username, "Str0ng#Pass1");
    const expense = (await createExpense(accessToken, {
      amount: "25.00",
      currency: "USD",
      category: "transport",
      description,
      date: "2025-01-15",
      type: "expense",
    })) as { id: string };
    testState.bobExpenseId = expense.id;
  },
);

Given("a user {string} has created an entry with an attachment", async ({}, username: string) => {
  await registerUser(username, `${username}@example.com`, "Str0ng#Pass1");
  const { accessToken } = await loginUser(username, "Str0ng#Pass1");
  const expense = (await createExpense(accessToken, {
    amount: "25.00",
    currency: "USD",
    category: "transport",
    description: "Bob's entry",
    date: "2025-01-15",
    type: "expense",
  })) as { id: string };
  // Upload an attachment via API so the scenario has an attachment to test against
  await uploadAttachmentApi(accessToken, expense.id, getReceiptJpgPath(), "receipt.jpg", "image/jpeg");
  testState.bobExpenseId = expense.id;
});

// ---------------------------------------------------------------------------
// When: UI-driven uploads — use waitForResponse for success paths
// ---------------------------------------------------------------------------

When("{word} uploads file {string} as an image attachment", async ({ page }, _username: string, _filename: string) => {
  const filePath = getReceiptJpgPath();
  const fileInput = page.locator('input[type="file"]');
  await Promise.all([
    page.waitForResponse(
      (r) => r.url().includes("/attachments") && r.request().method() === "POST",
      { timeout: 15000 },
    ),
    fileInput.setInputFiles(filePath),
  ]);
  await page.waitForLoadState("networkidle").catch(() => {});
});

When(
  "{word} uploads file {string} as a document attachment",
  async ({ page }, _username: string, _filename: string) => {
    const filePath = getInvoicePdfPath();
    const fileInput = page.locator('input[type="file"]');
    await Promise.all([
      page.waitForResponse(
        (r) => r.url().includes("/attachments") && r.request().method() === "POST",
        { timeout: 15000 },
      ),
      fileInput.setInputFiles(filePath),
    ]);
    await page.waitForLoadState("networkidle").catch(() => {});
  },
);

// Client-side rejection — no network request, just wait briefly
When("{word} attempts to upload file {string}", async ({ page }, _username: string, _filename: string) => {
  const filePath = getMalwareExePath();
  const fileInput = page.locator('input[type="file"]');
  await fileInput.setInputFiles(filePath);
  await page.waitForTimeout(1000);
});

When("{word} attempts to upload an oversized file", async ({ page }) => {
  const filePath = getOversizedFilePath();
  const fileInput = page.locator('input[type="file"]');
  await fileInput.setInputFiles(filePath);
  await page.waitForTimeout(1000);
});

When(
  "{word} clicks the delete button on attachment {string}",
  async ({ page }, _username: string, filename: string) => {
    // Use the specific aria-label to avoid clicking the expense delete button
    await page.getByRole("button", { name: new RegExp(`delete attachment ${filename}`, "i") }).click();
  },
);

// ---------------------------------------------------------------------------
// Then
// ---------------------------------------------------------------------------

Then("the attachment list should contain {string}", async ({ page }, filename: string) => {
  await expect(page.getByText(filename)).toBeVisible({ timeout: 10000 });
});

Then("the attachment should display as type {string}", async ({ page }, mimeType: string) => {
  await expect(page.getByText(mimeType)).toBeVisible();
});

Then("the attachment list should contain {int} items", async ({ page }, count: number) => {
  const attachmentItems = page
    .getByTestId("attachment-item")
    .or(page.getByRole("listitem").filter({ hasText: /\.(jpg|jpeg|pdf)/i }));
  await expect(attachmentItems).toHaveCount(count);
});

Then("the attachment list should include {string}", async ({ page }, filename: string) => {
  await expect(page.getByText(filename)).toBeVisible();
});

Then("the attachment list should not contain {string}", async ({ page }, filename: string) => {
  await expect(page.getByText(filename)).not.toBeVisible();
});

Then("an error message about unsupported file type should be displayed", async ({ page }) => {
  await expect(page.getByText(/unsupported|invalid file type|not allowed/i)).toBeVisible();
});

Then("the attachment list should remain unchanged", async ({ page }) => {
  const attachmentItems = page
    .getByTestId("attachment-item")
    .or(page.getByRole("listitem").filter({ hasText: /\.(jpg|jpeg|pdf)/i }));
  const count = await attachmentItems.count();
  expect(count).toBeGreaterThanOrEqual(0);
});

Then("an error message about file size limit should be displayed", async ({ page }) => {
  await expect(page.getByText(/too large|size limit|file.*size|maximum/i)).toBeVisible();
});

Then("the upload attachment button should not be visible", async ({ page }) => {
  await expect(page.locator('input[type="file"]')).not.toBeVisible();
});

Then("an access denied message should be displayed", async ({ page }) => {
  await expect(
    page.getByText(/access denied|forbidden|not authorized|permission|not found|failed to load/i).first(),
  ).toBeVisible();
});

Then("the delete attachment button should not be visible", async ({ page }) => {
  await expect(
    page.getByRole("button", {
      name: /delete.*attachment|remove.*attachment/i,
    }),
  ).not.toBeVisible();
});

Then("an error message about attachment not found should be displayed", async ({ page }) => {
  await expect(page.getByText(/not found|no longer exists|deleted/i)).toBeVisible();
});
